Main.hs

```diff
@@ -2,32 +2,142 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Data.Either
 
-data Expr
+data Expr =
+  Num Double
+  | Sqrt Expr
+  | Add Expr Expr
+  | Sub Expr Expr
+  | Mul Expr Expr
+  | Div Expr Expr
+  | Pow Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num a) = show a
+  show (Sqrt a) = show "sqrt(" ++ show a ++ ")" 
+  show (Add x y) = "(" ++ show x ++ " + " ++ show y ++ ")"
+  show (Sub x y) = "(" ++ show x ++ " - " ++ show y ++ ")"
+  show (Mul x y) = "(" ++ show x ++ " * " ++ show y ++ ")"
+  show (Div x y) = "(" ++ show x ++ " / " ++ show y ++ ")"
+  show (Pow x y) = "(" ++ show x ++ " ^ " ++ show y ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (==) (Num a) (Num b) = a == b
+  (==) (Sqrt a) (Sqrt b) = a == b
+  (==) (Add a b) (Add c d) = (a == c) && (b == d)
+  (==) (Sub a b) (Sub c d) = (a == c) && (b == d)
+  (==) (Mul a b) (Mul c d) = (a == c) && (b == d)
+  (==) (Div a b) (Div c d) = (a == c) && (b == d)
+  (==) (Pow a b) (Pow c d) = (a == c) && (b == d)
+  (==) _ _ = False
```

Comment:
Whenever your implementation of a derivable type class is default, use `deriving` instead.